File: src/main/java/one/pkg/modpublish/api/CurseForgeAPI.java
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package one.pkg.modpublish.api;

import com.google.gson.JsonObject;
import com.intellij.openapi.project.Project;
import okhttp3.*;
import one.pkg.modpublish.data.internel.ModInfo;
import one.pkg.modpublish.data.internel.PublishData;
import one.pkg.modpublish.data.local.DependencyInfo;
import one.pkg.modpublish.data.local.LauncherInfo;
import one.pkg.modpublish.data.local.MinecraftVersion;
import one.pkg.modpublish.data.network.curseforge.CurseForgeData;
import one.pkg.modpublish.data.network.curseforge.CurseForgePublishResult;
import one.pkg.modpublish.data.network.curseforge.ProjectRelation;
import one.pkg.modpublish.data.result.BackResult;
import one.pkg.modpublish.data.result.PublishResult;
import one.pkg.modpublish.data.result.Result;
import one.pkg.modpublish.settings.properties.PID;
import one.pkg.modpublish.util.io.JsonParser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.util.Optional;

public class CurseForgeAPI extends API {
    private static final String A_URL = "https://minecraft.curseforge.com/api/";
    private static final String B_URL = "https://api.curseforge.com/v1/";
    private boolean ab = false;

    @Override
    public @NotNull String getID() {
        return "CurseForge";
    }

    @Override
    public void updateABServer() {
        ab = !ab;
    }

    @Override
    public boolean getABServer() {
        return ab;
    }

    private Result create(PublishData data, Project project, File file, @Nullable BackResult bResult) {
        String modid = PID.CurseForgeModID.get(project);
        Request.Builder requestBuilder = getFormRequest(getRequestBuilder("projects/" + modid + "/upload-file", project));
        RequestBody fileData = RequestBody.create(file, MediaType.get("application/java-archive"));

        MultipartBody body = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("file", file.getName(), fileData)
                .addFormDataPart("metadata", createJsonBody(data, bResult))
                .build();
        Request request = requestBuilder.post(body).build();

        try (Response resp = NetworkUtil.client.newCall(request).execute()) {
            Optional<String> status = getStatus(resp);
            if (status.isPresent()) return PublishResult.create(this, status.get());
            String bs = resp.body().string();
            CurseForgePublishResult result = JsonParser.fromJson(bs, CurseForgePublishResult.class);
            if (result != null && result.isSuccess())
                return BackResult.result(result);
            return PublishResult.create(bs);
        } catch (IOException e) {
            return PublishResult.create(this, e.getMessage());
        }
    }

    @Override
    public @NotNull PublishResult createVersion(@NotNull PublishData data, @NotNull Project project) {
        if (ab) ab = false;
        BackResult bResult = null;
        for (int i = 0; i < data.files().length; i++) {
            File f = data.files()[i];
            var result = create(data, project, f, bResult);
            if (i == 0 && result instanceof BackResult br) bResult = br;
            if (result instanceof PublishResult pr && pr.isFailure()) return pr;
        }
        return PublishResult.empty();
    }

    @Override
    @NotNull
    String createJsonBody(@NotNull PublishData data, @NotNull Project project) {
        return createJsonBody(data, (BackResult) null);
    }

    @NotNull
    String createJsonBody(@NotNull PublishData data, @Nullable BackResult bResult) {
        CurseForgeData.CurseForgeDataBuilder builder = CurseForgeData.builder().releaseType(data.releaseChannel())
                .markdownChangelog(data.changelog())
                .displayName(data.versionName());
        if (bResult != null)
            builder.parentFileID(bResult.asCurseForgePublishResult().getId());
        for (MinecraftVersion v : data.minecraftVersions()) builder.gameVersion(v);
        if (data.supportedInfo().getClient().isEnabled())
            builder.gameVersion(data.supportedInfo().getClient().getCfid());
        if (data.supportedInfo().getServer().isEnabled())
            builder.gameVersion(data.supportedInfo().getServer().getCfid());
        for (LauncherInfo l : data.loaders()) if (l.getCfid() > 0) builder.gameVersion(l.getCfid());
        for (DependencyInfo d : data.dependencies()) {
            ModInfo info = d.getCurseforgeModInfo();
            if (info == null || info.modid() == null || info.modid().isBlank() ||
                    info.slug() == null || info.slug().isBlank()) continue;
            ProjectRelation relation = ProjectRelation.create(info.slug(), Integer.parseInt(info.modid()), d.getType());
            builder.dependency(relation);
        }
        return builder.build().toJson();
    }

    @Override
    public @NotNull ModInfo getModInfo(@NotNull String modid, @NotNull Project project) {
        if (!ab) ab = true;
        Request req = getJsonRequest(getRequestBuilder("mods/" + modid, project)).get().build();
        try (Response resp = NetworkUtil.client.newCall(req).execute()) {
            Optional<String> status = getStatus(resp);
            if (status.isPresent()) return ModInfo.of(status.get());
            JsonObject object = JsonParser.getJsonObject(resp.body().byteStream());
            JsonObject data = object.getAsJsonObject("data");
            return ModInfo.of(modid, data.get("name").getAsString(), data.get("slug").getAsString());
        } catch (IOException e) {
            return ModInfo.of(e.getMessage());
        }
    }

    @NotNull
    Request.Builder getRequestBuilder(@NotNull String url, @NotNull Project project) {
        Request.Builder builder = getBaseRequestBuilder();
        return ab ? builder.header("x-api-key",
                        PID.CurseForgeStudioToken.getProtect(project).data())
                .url(B_URL + url)
                : builder.header("X-Api-Token",
                        PID.CurseForgeToken.getProtect(project).data())
                .url(A_URL + url);
    }
}

File: src/main/kotlin/one/pkg/modpublish/api/CurseForgeAPI.kt
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package one.pkg.modpublish.api

import com.intellij.openapi.project.Project
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import one.pkg.modpublish.api.NetworkUtil.client
import one.pkg.modpublish.data.internal.ModInfo
import one.pkg.modpublish.data.internal.PublishData
import one.pkg.modpublish.data.network.curseforge.CurseForgeData
import one.pkg.modpublish.data.network.curseforge.CurseForgePublishResult
import one.pkg.modpublish.data.network.curseforge.ProjectRelation
import one.pkg.modpublish.data.result.BackResult
import one.pkg.modpublish.data.result.PublishResult
import one.pkg.modpublish.data.result.Result
import one.pkg.modpublish.settings.properties.PID
import one.pkg.modpublish.util.io.JsonParser.fromJson
import one.pkg.modpublish.util.io.JsonParser.getJsonObject
import one.pkg.modpublish.util.io.JsonParser.toJson
import java.io.File
import java.io.IOException

class CurseForgeAPI : API() {
    private var aBServer = false

    override val id: String
        get() = "CurseForge"

    override fun getAB(): Boolean = aBServer

    override fun updateAB() {
        aBServer = !aBServer
    }

    private fun create(data: PublishData, project: Project, file: File, bResult: BackResult?): Result {
        val modid = PID.CurseForgeModID.get(project)
        val requestBuilder = getFormRequest(getRequestBuilder("projects/$modid/upload-file", project))
        val fileBody = file.asRequestBody("application/java-archive".toMediaType())

        val body = MultipartBody.Builder().apply {
            setType(MultipartBody.FORM)
            addFormDataPart("file", file.name, fileBody)
            addFormDataPart("metadata", createJsonBody(data, bResult))
        }.build()

        val request = requestBuilder.post(body).build()

        return try {
            client.newCall(request).execute().use { resp ->
                getStatus(resp)?.let { return PublishResult.create(this, it) }
                val result = resp.body.string().fromJson(CurseForgePublishResult::class.java)
                if (result.isSuccess) BackResult.result(result) else PublishResult.create(resp.body.string())
            }
        } catch (e: IOException) {
            PublishResult.create(this, e.message)
        }
    }

    override fun createVersion(data: PublishData, project: Project): PublishResult {
        if (aBServer) aBServer = false
        var bResult: BackResult? = null
        data.files.forEachIndexed { index, file ->
            when (val result = create(data, project, file, bResult)) {
                is BackResult -> if (index == 0) bResult = result
                is PublishResult -> if (result.isFailure) return result
            }
        }
        return PublishResult.EMPTY
    }

    override fun createJsonBody(data: PublishData, project: Project): String =
        createJsonBody(data, null)

    fun createJsonBody(data: PublishData, bResult: BackResult?): String {
        return CurseForgeData().apply {
            this.releaseType(data.releaseChannel)
            this.markdownChangelog(data.changelog)
            this.displayName = data.versionName
            bResult?.let { this.parentFileID = it.asCurseForgePublishResult().id } ?: run {
                data.minecraftVersions.forEach { gameVersion(it) }
                if (data.supportedInfo.client.enabled) this.gameVersion(data.supportedInfo.client.cfid)
                if (data.supportedInfo.server.enabled) this.gameVersion(data.supportedInfo.server.cfid)
                data.loaders.filter { it.curseForgeVersion > 0 }.forEach { this.gameVersion(it.curseForgeVersion) }
            }
            data.dependencies.forEach { dep ->
                dep.curseforgeModInfo?.let { info ->
                    if (!info.modid.isNullOrBlank() && !info.slug.isNullOrBlank()) {
                        this.dependency(ProjectRelation.create(info.slug, info.modid.toInt(), dep.type))
                    }
                }
            }
        }.toJson()
    }

    override fun getModInfo(modid: String, project: Project): ModInfo {
        if (!aBServer) aBServer = true
        val request = getJsonRequest(getRequestBuilder("mods/$modid", project)).get().build()
        return try {
            client.newCall(request).execute().use { resp ->
                getStatus(resp)?.let { return ModInfo.of(it) }
                val data = resp.body.byteStream().getJsonObject().getAsJsonObject("data")
                ModInfo.of(modid, data.get("name").asString, data.get("slug").asString)
            }
        } catch (e: IOException) {
            ModInfo.of(e.message)
        }
    }

    fun getRequestBuilder(url: String, project: Project): Request.Builder =
        baseRequestBuilder.apply {
            if (aBServer) {
                header("x-api-key", PID.CurseForgeStudioToken.getProtect(project).data)
                url(B_URL + url)
            } else {
                header("X-Api-Token", PID.CurseForgeToken.getProtect(project).data)
                url(A_URL + url)
            }
        }

    companion object {
        private const val A_URL = "https://minecraft.curseforge.com/api/"
        private const val B_URL = "https://api.curseforge.com/v1/"
    }
}
